src/mono/msbuild/AotCompilerTask/MonoAotCompiler.cs

```diff
@@ -0,0 +1,249 @@
+    /// </summary>
+    [Required]
+    public ITaskItem[] Assemblies { get; set; } = Array.Empty<ITaskItem>();
+
+    /// <summary>
+    /// Assemblies which were AOT compiled.
+    ///
+    /// Successful AOT compilation will set the following metadata on the items:
+    ///   - AssemblerFile (when using OutputType=AsmOnly)
+    ///   - ObjectFile (when using OutputType=Normal)
+    ///   - AotDataFile
+    ///   - LlvmObjectFile (if using LLVM)
+    ///   - LlvmBitcodeFile (if using LLVM-only)
+    /// </summary>
+    [Output]
+    public ITaskItem[]? CompiledAssemblies { get; set; }
+
+    /// <summary>
+    /// Disable parallel AOT compilation
+    /// </summary>
+    public bool DisableParallelAot { get; set; }
+
+    /// <summary>
+    /// Use LLVM for AOT compilation.
+    /// The cross-compiler must be built with LLVM support
+    /// </summary>
+    public bool UseLlvm { get; set; }
+
+    /// <summary>
+    /// Choose between 'Normal', 'Full', 'LlvmOnly'.
+    /// LlvmOnly means to use only LLVM for FullAOT, AOT result will be a LLVM Bitcode file (the cross-compiler must be built with LLVM support)
+    /// </summary>
+    public string AotMode { get; set; } = nameof(MonoAotMode.Normal);
+
+    /// <summary>
+    /// Choose between 'Normal', 'AsmOnly'
+    /// AsmOnly means the AOT compiler will produce .s assembly code instead of an .o object file.
+    /// </summary>
+    public string OutputType { get; set; } = nameof(MonoAotOutputType.Normal);
+
+    /// <summary>
+    /// Path to LLVM binaries (opt and llc)
+    /// It's required if UseLlvm is set
+    /// </summary>
+    public string? LlvmPath { get; set; }
+
+    /// <summary>
+    /// Path to the directory where msym artifacts are stored.
+    /// </summary>
+    public string? MsymPath { get; set; }
+
+    ConcurrentBag<ITaskItem> compiledAssemblies = new ConcurrentBag<ITaskItem>();
+    MonoAotMode parsedAotMode;
+    MonoAotOutputType parsedOutputType;
+
+    public override bool Execute()
+    {
+        Utils.Logger = Log;
+
+        if (string.IsNullOrEmpty(AotCompilerBinary))
+        {
+            throw new ArgumentException($"'{nameof(AotCompilerBinary)}' is required.", nameof(AotCompilerBinary));
+        }
+
+        if (!File.Exists(AotCompilerBinary))
+        {
+            throw new ArgumentException($"'{AotCompilerBinary}' doesn't exist.", nameof(AotCompilerBinary));
+        }
+
+        if (Assemblies.Length == 0)
+        {
+            throw new ArgumentException($"'{nameof(Assemblies)}' is required.", nameof(Assemblies));
+        }
+
+        if (UseLlvm && string.IsNullOrEmpty(LlvmPath))
+        {
+            // prevent using some random llc/opt from PATH (installed with clang)
+            throw new ArgumentException($"'{nameof(LlvmPath)}' is required when '{nameof(UseLlvm)}' is true.", nameof(LlvmPath));
+        }
+
+        if (!Enum.TryParse<MonoAotMode>(AotMode, out parsedAotMode))
```

Comment:
Not a fan of leaking enum members names. I'd just make a switch to make it obvious